docs/task_model.md

```diff
@@ -0,0 +1,173 @@
+most important field inside `status` is `state`, which indicates where the task
+is in its lifecycle (assigned, running, complete, and so on). The status
+information in this field may become out of date if the node that the task is
+assigned to is unresponsive. In this case, it's up to the orchestrator to
+replace the task with a new one.
+
+### Desired state
+
+Desired state is the state that the orchestrator would like the task to progress
+to. This field provides a way for the orchestrator to control when the task can
+advance in state. For example, the orchestrator may create a task with desired
+state set to `READY` during a rolling update, and then advance the desired state
+to `RUNNING` once the old task it is replacing has stopped. This gives it a way
+to get the new task ready to start (for example, pulling the new image), without
+actually starting it.
+
+## Properties of tasks
+
+A task is a "one-shot" execution unit. Once a task stops running, it is never
+executed again. A new task may be created to replace it.
+
+Tasks states are changed in a monotonic progression. Tasks may move to states
+beyond the current state, but their states may never move backwards.
+
+## Task history
+
+Once a task stops running, the task object is not necessarily removed from the
+distributed data store. Generally, a few historic tasks for each slot of each
+service are retained to provide task history. The task reaper will garbage
+collect old tasks if the limit of historic tasks for a given slot is reached.
+Currently, retention of containers on the workers is tied to the presence of the
+old task objects in the distributed data store, but this may change in the
+future.
+
+## Task lifecycle
+
+Tasks are created by the orchestrator. They may be created for a new service, or
+to scale up an existing service, or to replace tasks for an existing service
+that are no longer running for whatever reason. The orchestrator creates tasks
+in the `NEW` state.
+
+Tasks next run through the allocator, which allocate resources such as network
+attachments which are necessary for the tasks to run. When the allocator has
+processed a task, it moves the task to the `ALLOCATED` state.
+
+The scheduler takes `ALLOCATED` tasks and assigns them to nodes (or verifies
+that the requested node has the necessary resources, in the case of global
+services' tasks). It changes their state to `ASSIGNED`.
+
+From this point, control over the state passes to the agent. A task will
+progress through the `ACCEPTED`, `PREPARING`, and `STARTING` states on the
+way to `RUNNING`. If a task exits without an error code, it moves to the
+`COMPLETE` state. If it fails, it moves to the `FAILED` state instead.
+
+A task may alternatively end up in the `SHUTDOWN` state if its shutdown was
+requested by the orchestrator, or the `REJECTED` state if the agent rejected the
+task.
+
+## Slot model
+
+Replicated tasks have a slot number assigned to them. This allows the system to
```

Comment:
I still don't understand why we don't extend the slot model to global tasks, as well. In the case of global tasks, there is a slot per node.